### PR TITLE
chore: migrate from kuttl to chainsaw

### PR DESCRIPTION
**Pull request by eddycharly**
This PR migrates e2e tests from kuttl to chainsaw.

Chainsaw was created because kuttl stopped to be actively maintained (it's actually a single maintainer now).

We are also solving most of the kuttl limitations that exist today and wrote a command to automatically migrate from kuttl to chainsaw. It supercharges e2e tests with better logs, config maps assertions, [assertions trees](https://kyverno.io/blog/2023/12/13/kyverno-chainsaw-exploring-the-power-of-assertion-trees/) and many more things 💪

This is a relatively young project but it has already been adopted by a couple of e2e tests heavy projects (kyverno, keptn, fairwinds rbac-manager, grafana-operator, redis-operator and others).

To give more context, you can see similar PRs at:
- https://github.com/keptn/lifecycle-toolkit/pull/3000
- https://github.com/operator-framework/operator-sdk/pull/6676
- https://github.com/grafana/tempo-operator/pull/786 in this PR the tempo-operator maintainers asked some good questions and I explained the current status, that I suggest taking a look at.

Sadly, [kuttl](https://github.com/kudobuilder/kuttl) haven't done a release since jan 2023.

Here you can find a good video showcasing chainsaw: https://www.youtube.com/watch?v=Ejof-wtAdQM&ab_channel=CamilaMacedo

To make things more concrete, there are a couple of issues in this repo, a few files are not valid and kuttl doesn't complain because unmarshalling is not strict so the non-valid fields are simply ignored. This is a problem because the behaviour of the tests is not the expected one, for example:
- https://github.com/open-telemetry/opentelemetry-operator/blob/c534eafdf30e7bd8c484ea70ffc7d5a6a9294e41/tests/e2e-opampbridge/opampbridge/00-assert.yaml#L67-L70
- https://github.com/open-telemetry/opentelemetry-operator/blob/94c8420ea6403ab359af9eb5c20ddef8d243a9d2/tests/e2e-prometheuscr/create-sm-prometheus-exporters/07-delete.yaml#L6-L8

If you need more infos :point_down:
- GitHub: https://github.com/kyverno/chainsaw
- Docs: https://kyverno.github.io/chainsaw
- Slack: https://kubernetes.slack.com/archives/C067LUFL43U

### CODE REVIEW

**Comment by iblancasa** [comment]
cc/ @IshwarKanse

File: Makefile [hunk 1]
```diff
@@ -246,6 +249,28 @@ e2e-multi-instrumentation:
...
 	$(KUTTL) test --config kuttl-test-opampbridge.yaml
 
+############
+# CHAINSAW #
+############
+
+# end-to-end-test for testing pdb support
+.PHONY: chainsaw-e2e-pdb
+chainsaw-e2e-pdb:
+	kubectl apply -f ./tests/_build/artifacts/
```
Comment (eddycharly) [inline Makefile:259]: Where is this built ?

File: .github/workflows/e2e.yaml [hunk 2]
```diff
@@ -42,33 +41,38 @@ jobs:
...
         path: bin
         key: ${{ runner.os }}-${{ runner.arch }}-${{ hashFiles('Makefile') }}
-
     - name: Install tools
       run: make install-tools
-
-    - name: "run tests"
+    - name: Install chainsaw
+      uses: kyverno/action-install-chainsaw@07b6c986572f2abaf6647c85d37cbecfddc4a6ab # v0.1.3
+    - name: Prepare e2e tests
```
Comment (swiatekm-sumo) [inline .github/workflows/e2e.yaml:57]: I would prefer this to be part of `make install-tools`, and for that target to install chainsaw locally just like we do for every other tool necessary to run e2e tests.

File: Makefile [hunk 3]
```diff
@@ -371,10 +366,9 @@ KUSTOMIZE_VERSION ?= v5.0.3
 CONTROLLER_TOOLS_VERSION ?= v0.12.0
 GOLANGCI_LINT_VERSION ?= v1.54.0
 KIND_VERSION ?= v0.20.0
-KUTTL_VERSION ?= 0.15.0
 
 .PHONY: install-tools
-install-tools: kustomize golangci-lint kind controller-gen envtest crdoc kuttl kind operator-sdk
```
Comment (swiatekm-sumo) [inline Makefile:377]: Can we not install chainsaw locally the same way we did for kuttl?

**Review by swiatekm-sumo — COMMENTED** [review]
Some comments on managing chainsaw in a local environment.

File: Makefile (hunk 3 above)
Comment (eddycharly) [reply Makefile:377]: Sure we can do that. What we did in grafana-operator was to support both, only installing it if not already installed.

Comment (swiatekm-sumo) [reply Makefile:377]: Sure, that works as well.

File: .github/workflows/e2e.yaml (hunk 2 above)
Comment (swiatekm-sumo) [reply .github/workflows/e2e.yaml:57]: Can dependabot upgrade actions pinned to a specific commit?

File: Makefile (hunk 3 above)
Comment (eddycharly) [reply Makefile:377]: Ok, i'll do that.

File: .github/workflows/e2e.yaml (hunk 2 above)
Comment (eddycharly) [reply .github/workflows/e2e.yaml:57]: Yup, works perfectly fine.

Comment (eddycharly) [reply .github/workflows/e2e.yaml:57]: See https://github.com/kyverno/kyverno/pull/9715/files for example.

Comment (eddycharly) [reply .github/workflows/e2e.yaml]: Done.

**Comment by eddycharly** [comment]
Quick blog post covering chainsaw two major new features in `v0.1.4` https://kyverno.io/blog/2024/02/15/kyverno-chainsaw-0.1.4-awesome-new-features/ 🤩

**Comment by IshwarKanse** [comment]
@eddycharly The work done on Chainsaw is really impressive. I'm looking forward to use it for other operator testing as well and for our internal QE testing.

**Comment by eddycharly** [comment]
> @eddycharly The work done on Chainsaw is really impressive. I'm looking forward to use it for other operator testing as well and for our internal QE testing.

@IshwarKanse glad to know you like it !

**Comment by eddycharly** [comment]
Not sure what breaks here, probably something simple but i fail to spot it 🙈

**Comment by eddycharly** [comment]
Hope i fixed it now 🤞

**Pull request closed without merging**